File: c++/0664.cpp
// Link: https://leetcode.com/problems/strange-printer/description/
// Runtime: 128 ms
// Memory: 9.4 MB

class Solution 
{
public:
    int strangePrinter(const std::string& s) 
    {
        std::vector<std::vector<int>> segmentsChanging(s.size(), std::vector<int>(s.size(), s.size()));
        for (int length = 1; length <= static_cast<int>(s.size()); length++)
        {
            for (int left = 0; left <= static_cast<int>(s.size() - length); left++)
            {
                int right = left + length - 1;
                int j = -1;
                for (int i = left; i < right; i++)
                {
                    if ((s[i] != s[right]) && (j == -1)) j = i;

                    if (j != -1) 
                        segmentsChanging[left][right] = std::min(segmentsChanging[left][right], 1 + segmentsChanging[j][i] + segmentsChanging[i + 1][right]);
                }

                if (j == -1) segmentsChanging[left][right] = 0;
            }
        }

        return 1 + segmentsChanging[0][static_cast<int>(s.size() - 1)];
    }
};

File: python/0664.py
# Link: https://leetcode.com/problems/strange-printer/description/
# Runtime: 1374 ms
# Memory: 16.3 MB


class Solution:
    def strangePrinter(self, s: str) -> int:
        segments_changing = [[len(s)] * len(s) for _ in range(len(s))]
        for length in range(1, len(s) + 1):
            for left in range(len(s) - length + 1):
                right = left + length - 1
                j = -1
                for i in range(left, right):
                    if (s[i] != s[right]) and (j == -1):
                        j = i

                    if j != -1:
                        segments_changing[left][right] = min(
                            segments_changing[left][right],
                            1
                            + segments_changing[j][i]
                            + segments_changing[i + 1][right],
                        )

                if j == -1:
                    segments_changing[left][right] = 0

        return 1 + segments_changing[0][-1]
